File: apple/QuenderinKit/Sources/QuenderinKit/ChatModel.swift
import Foundation
import Combine

/// A document the user attached to a chat message — name for display, extracted text for the
/// model. Extraction happens AT ATTACH TIME (strict UTF-8, size-capped) so what the model sees
/// is fixed when the message is sent, even if the file changes later. Twin of Kotlin
/// `AttachedDocument` (Milestone 1, roadmap Stage 2 "documents in chat").
public struct AttachedDocument: Sendable, Equatable, Codable {
    public let name: String
    public let text: String

    public init(name: String, text: String) {
        self.name = name
        self.text = text
    }
}

public struct ChatMessage: Sendable, Identifiable, Equatable {
    public enum Role: String, Sendable { case user, assistant }
    public let id: UUID
    public let role: Role
    public var text: String
    /// Documents attached to this (user) message. The transcript shows chips + the typed text;
    /// `engineText` is what the model gets.
    public var documents: [AttachedDocument]

    public init(id: UUID = UUID(), role: Role, text: String, documents: [AttachedDocument] = []) {
        self.id = id
        self.role = role
        self.text = text
        self.documents = documents
    }

    /// The text the ENGINE sees: attached documents first (clearly labeled), then the typed
    /// message. Kept out of `text` so bubbles/persistence previews stay readable, but composed
    /// into every windowed history pass — follow-up questions still have the document in context.
    public var engineText: String {
        guard !documents.isEmpty else { return text }
        let docs = documents.map { "Attached file \"\($0.name)\":\n\($0.text)" }.joined(separator: "\n\n")
        return "\(docs)\n\n\(text)"
    }
}

public extension ChatMessage {
    /// True when this is an assistant message whose text trips `SafetyBlocklist` — the chat UI
    /// surfaces a non-blocking warning (`SupportContact.flaggedOutputNotice`) rather than
    /// suppressing it, the on-device "minimize risk" safeguard for the Generative-AI policies.
    /// User messages are never flagged. Kept in parity with Android `ChatMessage.isFlagged`.
    var isFlagged: Bool { role == .assistant && SafetyBlocklist.isBlocked(text) }
}

/// Drives a chat conversation against any `InferenceEngine`, appending tokens to
/// the in-flight assistant message as they stream in. Runs on the mock engine
/// today; swaps to `LlamaEngine` with no change.
@MainActor
public final class ChatModel: ObservableObject {
    @Published public private(set) var messages: [ChatMessage] = []
    @Published public private(set) var isGenerating = false
    /// True when the last settled reply stopped because it hit `options.maxTokens` (not Stop, not
    /// EOG). The chat UI surfaces a "Continue" chip so a mid-sentence cut is recoverable
    /// (KNOWN_FAILURE_MODES). Cleared on the next send / reset / restore.
    @Published public private(set) var lastHitTokenCap = false

    /// Set by `stopGenerating()`; checked per token so a stop lands within one token.
    private var stopRequested = false

    /// Stop the current reply where it is — the streamed partial stays in the transcript
    /// (stopping is a decision about YOUR time, not a failure; no error styling).
    public func stopGenerating() {
        stopRequested = true
        // The token loop below only lands a stop AT a token boundary, and never during prefill
        // (before the first token) — so also tell the engine to interrupt its native decode now,
        // ending generation in <500ms even mid-prefill (Q-005/Q-217).
        engine.requestCancel()
    }

    /// Extend the previous reply after a token-cap stop. Sends a short continue cue so the model
    /// picks up mid-thought. No-op when `lastHitTokenCap` is false.
    public func continueLast(options: GenerationOptions = .init()) async {
        guard lastHitTokenCap, !isGenerating else { return }
        await send("Continue from where you left off. Do not repeat what you already wrote.", options: options)
    }

    private let engine: InferenceEngine
    private let context: ConversationContext

    public init(engine: InferenceEngine, context: ConversationContext = .init()) {
        self.engine = engine
        self.context = context
    }

    /// Send a prompt and stream the reply. No-ops on empty input (unless documents are
    /// attached — "summarize this file" with no extra words is a legitimate send) or while a
    /// previous generation is still running.
    public func send(_ prompt: String, documents: [AttachedDocument] = [], options: GenerationOptions = .init()) async {
        let trimmed = prompt.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty || !documents.isEmpty, !isGenerating else { return }

        isGenerating = true
        stopRequested = false
        lastHitTokenCap = false
        defer { isGenerating = false }

        // The engine's REAL loaded `n_ctx` (often 512–2048 on phones), so the history trim below
        // matches the actual native window instead of the configured 4096 (Q-167 — Android had this,
        // iOS didn't; twin-drift audit). Awaited BEFORE any transcript mutation and only AFTER
        // `isGenerating` is set: every await yields the main actor, and a reentrant send must not
        // slip past the guard (the @MainActor+await journal pattern).
        let realContextTokens = await engine.loadedContextTokens()

        messages.append(ChatMessage(role: .user, text: trimmed, documents: documents))
        // Chat-structured generation: pass the budget-windowed history so the engine formats it
        // with the model's OWN chat template (answers as an assistant, stops at end-of-turn).
        // Template-less engines fall back to the flat transcript; the assistant remembers prior
        // turns either way. Attached documents are composed in (engineText) BEFORE windowing so
        // the token budget counts them.
        let windowed = context.windowedHistory(messages.map { message in
            guard !message.documents.isEmpty else { return message }
            var composed = message
            composed.text = message.engineText
            composed.documents = []
            return composed
        }, contextTokensOverride: realContextTokens)
        var assistant = ChatMessage(role: .assistant, text: "")
        messages.append(assistant)
        let assistantID = assistant.id   // track by id, NOT a captured index — `messages` can be mutated

        var tokenCount = 0
        var hitDegeneration = false
        do {
            let stream = try await engine.generateChat(system: context.systemPrompt, history: windowed, options: options)
            for try await token in stream {
                // Dropping the iterator on break terminates the stream (the engine's
                // onTermination stops decoding), so Stop also stops the compute.
                if stopRequested { break }
                // Degeneration guard: if the tail is verbatim-looping despite the sampler's
                // repetition penalty, stop paying for tokens — the collapse below cleans up.
                tokenCount += 1
                if tokenCount % 32 == 0, DegenerationGuard.looksDegenerate(assistant.text) {
                    hitDegeneration = true
                    break
                }
                assistant.text += token
                // `send` is @MainActor, but every `await` above yields the actor — so `reset()`
                // (clear) or `restore()` (open history) can mutate `messages` mid-stream. Look the
                // message up by id each token: a captured index would crash (out of range) or write
                // to the wrong message. If it's gone, the user moved on — stop streaming into it.
                guard let i = messages.firstIndex(where: { $0.id == assistantID }) else { return }
                messages[i] = assistant
            }
        } catch {
            assistant.text = "⚠️ " + OnboardingModel.describe(error)
            if let i = messages.firstIndex(where: { $0.id == assistantID }) { messages[i] = assistant }
            return
        }
        // Settle: collapse any exact-duplicate paragraph runs that slipped through (covers
        // the guard's between-checkpoints window and the Stop path alike), and never leave a
        // silently EMPTY bubble — an engine that produced zero tokens gets an honest notice.
        assistant.text = DegenerationGuard.collapseRepeatedParagraphs(assistant.text)
        if assistant.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty, !stopRequested {
            assistant.text = "The model returned an empty reply. Try rephrasing, or pick a larger model in the Model library."
        }
        // Token-cap mid-sentence → Continue. Count is per streamed piece from the UTF-8 decoder
        // (≈ per token on iOS). Not Stop, not degeneration, and we hit the budget.
        if !stopRequested, !hitDegeneration, tokenCount >= options.maxTokens {
            lastHitTokenCap = true
        }
        if let i = messages.firstIndex(where: { $0.id == assistantID }) { messages[i] = assistant }
    }

    /// Clear the conversation.
    public func reset() {
        // Q-322: cancel any in-flight decode BEFORE wiping its target array — otherwise the streaming
        // loop keeps appending tokens into a transcript that's just been replaced (cross-chat bleed).
        engine.requestCancel()
        lastHitTokenCap = false
        messages.removeAll()
    }

    /// Replace the transcript with a previously persisted conversation (see `ConversationStore`),
    /// so a chat picks up exactly where it left off after a relaunch.
    public func restore(_ saved: [ChatMessage]) {
        engine.requestCancel()   // Q-323: same — stop the decode before swapping the transcript out
        lastHitTokenCap = false
        messages = saved
    }
}

File: android/quenderin-core/src/main/kotlin/ai/quenderin/core/ChatModel.kt
package ai.quenderin.core

enum class Role { USER, ASSISTANT }

/**
 * A document the user attached to a chat message — name for display, extracted text for the
 * model. Extraction happens AT ATTACH TIME (strict UTF-8, size-capped) so what the model sees is
 * fixed when the message is sent. Twin of iOS `AttachedDocument` (Milestone 1).
 */
data class AttachedDocument(val name: String, val text: String)

data class ChatMessage(
    val role: Role,
    val text: String,
    /** Documents attached to this (user) message. UI shows chips + typed text; [engineText] is
     *  what the model gets. */
    val documents: List<AttachedDocument> = emptyList(),
) {
    /** The text the ENGINE sees: labeled documents first, then the typed message. Composed into
     *  every windowed history pass so follow-ups keep the document in context. Twin of iOS. */
    val engineText: String
        get() {
            if (documents.isEmpty()) return text
            val docs = documents.joinToString("\n\n") { "Attached file \"${it.name}\":\n${it.text}" }
            return "$docs\n\n$text"
        }
}

/**
 * True when this is an assistant message whose text trips [SafetyBlocklist] — the chat UI surfaces
 * a non-blocking warning ([SupportContact.FLAGGED_OUTPUT_NOTICE]) rather than suppressing it, the
 * on-device "minimize risk" safeguard for the Generative-AI policies. User messages are never
 * flagged. Kept in parity with iOS `ChatMessage.isFlagged`.
 */
val ChatMessage.isFlagged: Boolean
    get() = role == Role.ASSISTANT && SafetyBlocklist.isBlocked(text)

/**
 * M2 chat brain: holds the transcript, sends a prompt to the [engine], appends the
 * reply. Dependency-free and listener-based ([onChange]) so it unit-tests on the JVM;
 * the Compose layer maps [onChange] into state and runs [send] off the main thread
 * (optionally via [LlamaEngine.complete]'s streaming overload for token-by-token UI).
 * Twin of Swift `ChatModel`.
 *
 * Concurrency contract (Android runs [send] on a background dispatcher, unlike iOS's single
 * `@MainActor`): the transcript is only ever mutated under [lock], and each streaming generation
 * carries a monotonic id ([activeGeneration]). A lifecycle op that changes the transcript from the
 * main thread — [reset]/[restore], driven by conversation switch/new/delete — first [stopGenerating]s
 * the in-flight send, which bumps the generation id and cancels the native decode. A still-running
 * (zombie) send then finds its id superseded and drops every remaining token/settle write, so tokens
 * from conversation A can never bleed into (or index out of bounds on) conversation B. `@Volatile` +
 * `synchronized(lock)` give the Compose main-thread reader a happens-before view of the writes, same
 * idiom as [AgentSession]. (Q-004/Q-168 corruption; Q-005/Q-237 stop + degeneration abort.)
 */
class ChatModel(
    private val engine: InferenceEngine,
    var onChange: (List<ChatMessage>) -> Unit = {},
    private val context: ConversationContext = ConversationContext(),
) {
    private val lock = Any()
    private val _messages = mutableListOf<ChatMessage>()
    val messages: List<ChatMessage> get() = synchronized(lock) { _messages.toList() }

    /** Monotonic id of the generation allowed to write the transcript. A send captures its id at start;
     *  [stopGenerating]/[reset]/[restore] bump this so a superseded (zombie) send stops writing. */
    @Volatile
    private var activeGeneration: Long = 0

    /** True while a [send] is streaming. Read on the main thread (Compose) to drive the Stop affordance;
     *  written under [lock]. Twin of iOS `ChatModel.isGenerating`. */
    @Volatile
    var isGenerating: Boolean = false
        private set

    private fun emit() = onChange(messages)

    /**
     * Appends the user's line, runs the engine over the whole transcript streaming into a placeholder
     * assistant message, settles the final text, and returns it. Blocking; the app calls it off the main
     * thread. Throws [EngineNotLoadedException] if no model is loaded (surfaced to the UI, never
     * swallowed). No-op (returns "") if a generation is already in flight — a rapid double-send can't
     * race two writers onto the transcript.
     */
    @JvmOverloads
    fun send(text: String, documents: List<AttachedDocument> = emptyList()): String {
        val trimmed = text.trim()
        // Documents alone are a legitimate send ("summarize this file" with no extra words). Twin-drift
        // fix: an empty send is a SILENT no-op (return ""), matching iOS ChatModel's `guard … else { return }`
        // — the old `require(...)` threw IllegalArgumentException, crashing the send coroutine on Android
        // while iOS ignored it, and it was inconsistent even with this file's own already-generating guard
        // (which returns "" too).
        if (trimmed.isEmpty() && documents.isEmpty()) return ""
        // The user line + placeholder assistant slot + this generation's id are established atomically
        // under the lock, so a concurrent reset/restore either happens fully before (this send sees the
        // new transcript) or fully after (it bumps our id and we drop our writes below), never
        // interleaved. Two emits (user, then placeholder) keep the UX where the user's bubble appears
        // immediately, then the "typing…" placeholder — the snapshots are taken inside the lock and
        // emitted outside it (a listener must never re-enter under the lock).
        val myGen: Long
        val placeholderIndex: Int
        val history: List<ChatMessage>
        val afterUser: List<ChatMessage>
        val afterPlaceholder: List<ChatMessage>
        synchronized(lock) {
            if (isGenerating) return ""
            isGenerating = true
            myGen = ++activeGeneration
            _messages += ChatMessage(Role.USER, trimmed, documents)
            afterUser = _messages.toList()
            // History is computed BEFORE the placeholder is appended so the empty assistant turn isn't
            // included in the prompt. Passed STRUCTURED so the engine formats with the model's own chat
            // template (early-stop + higher quality). Trimmed to the engine's REAL loaded n_ctx (often
            // 512–2048 on phones), not a hardcoded 4096 that would overflow the native window (Q-167).
            // Attached documents are composed in (engineText) BEFORE windowing so the budget counts them.
            history = context.windowedHistory(
                _messages.map { if (it.documents.isEmpty()) it else it.copy(text = it.engineText, documents = emptyList()) },
                engine.loadedContextTokens,
            )
            placeholderIndex = _messages.size
            _messages += ChatMessage(Role.ASSISTANT, "")
            afterPlaceholder = _messages.toList()
        }
        onChange(afterUser)
        onChange(afterPlaceholder)
        try {
            val sb = StringBuilder()
            var tokenCount = 0
            // Stream into the placeholder so the reply appears token-by-token instead of the UI sitting
            // blank for the whole (multi-second) generation. Mirrors iOS `ChatModel`.
            val reply = engine.completeChat(context.systemPrompt, history) { piece ->
                sb.append(piece)
                // Degeneration guard: if the tail is verbatim-looping despite the sampler's repetition
                // penalty, stop paying for tokens — the settle-time collapse below cleans up. Checked every
                // 32 tokens like iOS; asks the engine to end the native decode now. (Q-237)
                tokenCount += 1
                if (tokenCount % 32 == 0 && DegenerationGuard.looksDegenerate(sb.toString())) {
                    engine.requestCancel()
                }
                // Drop the write if this generation was superseded (conversation switch/new/reset/restore,
                // or Stop) — a captured index would otherwise write into the WRONG transcript or throw out
                // of bounds. Twin of iOS's "look the message up by id each token; if it's gone, stop".
                writeAssistant(myGen, placeholderIndex, sb.toString())
            }
            // Settle on the authoritative full text (covers the non-streaming fallback, where onToken never
            // fired and the placeholder is still empty) — with duplicate-paragraph runs collapsed
            // (DegenerationGuard; the sampler penalty upstream prevents most, this cleans the rest).
            val settled = DegenerationGuard.collapseRepeatedParagraphs(reply)
            // Q-588: never leave a silently EMPTY assistant bubble — an engine that produced zero tokens gets
            // an honest notice, mirroring iOS ChatModel. Only when this generation is still active: a user
            // Stop supersedes it (activeGeneration bumped), and then the streamed partial should stand,
            // exactly as iOS's `!stopRequested` guard. (writeAssistant also drops a superseded write.)
            val stillActive = synchronized(lock) { myGen == activeGeneration }
            val finalText = if (settled.isBlank() && stillActive) {
                "The model returned an empty reply. Try rephrasing, or pick a larger model in the Model library."
            } else {
                settled
            }
            writeAssistant(myGen, placeholderIndex, finalText)
            return finalText
        } finally {
            synchronized(lock) {
                // Only the CURRENT generation owns isGenerating; a superseded send must not clear a flag a
                // newer send/lifecycle op already re-set (M10 finally-safety, same as AgentSession).
                if (activeGeneration == myGen) isGenerating = false
            }
        }
    }

    /** Write the assistant placeholder iff this generation is still the active one and the slot is still
     *  in range — under [lock] so a concurrent [reset]/[restore] can't shrink the list between the check
     *  and the write. A superseded generation (or a cleared slot) drops the write silently. */
    private fun writeAssistant(gen: Long, index: Int, textValue: String) {
        val changed = synchronized(lock) {
            if (gen != activeGeneration || index >= _messages.size) return@synchronized false
            _messages[index] = ChatMessage(Role.ASSISTANT, textValue)
            true
        }
        if (changed) emit()
    }

    /**
     * Stop the in-flight reply where it is: bump the generation id so no more tokens land, and ask the
     * engine to interrupt its native decode now (the flag-only break lands one token late and does nothing
     * during a non-interruptible prefill — Q-005/Q-217). Idempotent; no-op when nothing is generating.
     * The streamed partial already in the transcript stays. Twin of iOS `ChatModel.stopGenerating`.
     */
    fun stopGenerating() {
        synchronized(lock) {
            if (!isGenerating) return
            activeGeneration++   // supersede the running send → its remaining writes become no-ops
            isGenerating = false
        }
        engine.requestCancel()
    }

    /** Clear the conversation. First stops any in-flight generation so a zombie send can't write into the
     *  now-empty transcript (out of bounds) or resurrect a stale reply. */
    fun reset() {
        stopGenerating()
        synchronized(lock) { _messages.clear() }
        emit()
    }

    /** Replace the transcript with a previously persisted conversation (see [ConversationStore]),
     *  so a chat picks up exactly where it left off after a relaunch. First stops any in-flight
     *  generation so tokens from the OUTGOING conversation can't bleed into the one being restored
     *  (the conversation-switch corruption — Q-004/Q-168). */
    fun restore(saved: List<ChatMessage>) {
        stopGenerating()
        synchronized(lock) {
            _messages.clear()
            _messages.addAll(saved)
        }
        emit()
    }
}
